### DISCUSSION TITLE
CAN Bus data processing

### DISCUSSION THREAD

**Discussion by loki429** [lf-edge/ekuiper#2235] [Q&A] [answered]
Hello, I'm a newb to ekuiper and hoping someone can assist me with processing CAN bus data please?
I found this article that makes it sound really simple but I just can't get it to work :-(
https://www.emqx.com/en/blog/bridging-demanded-signals-from-can-bus-to-mqtt-by-ekuiper

From the blog: _"eKuiper understands CAN bud data. It eases the CAN frame decoding and transforms it into just some configurations. To consume CAN bus data, you can create a stream using the following SQL statement:"_
`CREATE STREAM canDemo () WITH (TYPE="can", FORMAT="can", SHARED="TRUE", SCHEMAID="dbc")`

Where do I add that? From the management gui there is no CAN type, and from the kuiper cli it doesn't recognise the type.
If I add it as SQL in a rule I receive an error saying the option 'format=can' is invalid.

Many thanks for any assistance or pointers in the right direction.
Jon

**Comment by ngjaying** [comment, marked as answer]
Hi Jon, 
The CAN source is not open source yet.  You can send me an email huangjy at emqx.io so that I can redirect you to the right person. 
Thank you.

**Reply by epvin** [reply]
I also want to know how to obtain the source of CAN bus. Could you please share it with me? My email is 506002470@qq.com Please kindly send the go file to my email. Thank you very much

**Reply by ngjaying** [reply]
EMQ provide a free trail docker image here https://hub.docker.com/r/emqx/ek-can. You can have a try and contact them if need supports.

**Reply by epvin** [reply]
Received your email, thank you very much for your patient help

**Reply by KuSunoKiAMe** [reply]
hello,I also need a copy of the source code too. Would it be convenient for you to read it for me? My email is j624878920@163.com.Thank you very much

**Reply by ngjaying** [reply]
Please check https://hub.docker.com/r/emqx/ek-can

**Reply by KuSunoKiAMe** [reply]
I have tried out this Docker image and completed an mqtt message transmission.But if I need to listen to multiple can ports, what should I do

**Reply by ngjaying** [reply]
Can you create multiple streams?

**Reply by tarmacx** [reply]
It's been a while since the last update of the ek-can. Is it still in development ? Any plans to open source it (you said "yet") ? Can the can component be split from the ek-can image so we can use a newer version with the closed source can component ?

**Reply by ngjaying** [reply]
@tarmacx There is no plan to open source currently. The can module and eKuiper are now part of neuronEX product. Check https://www.emqx.com/en/products/neuronex.

**Reply by tarmacx** [reply]
Good to know, will we still see some development to the open source ekuiper ? Is it worth it to build an CAN addon for ekuiper ? I had a look at the doc of NeuronEX and couldnt find CAN copmenents, am I missing something ? regards.

**Reply by ngjaying** [reply]
Open source eKuiper is actively developing. You can tell from the project commit history.  However, the CAN addon only plans to maintain in NeuronEX. The doc can be found here  [CAN source](https://docs.emqx.com/en/neuronex/latest/streaming-processing/can.html). If you are interested in it, you can contact the EMQX team from the website. Thanks.
